benches/main.rs

```diff
@@ -0,0 +1,106 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+
+use event_manager::utilities::subscribers::{
+    CounterInnerMutSubscriber, CounterSubscriber, CounterSubscriberWithData,
+};
+use event_manager::{EventManager, EventSubscriber, MutEventSubscriber, SubscriberOps};
+use std::sync::{Arc, Mutex};
+
+pub fn run_basic_subscriber(c: &mut Criterion) {
+    let mut event_manager = EventManager::<CounterSubscriber>::new().unwrap();
+    for _ in 0..200 {
+        let mut counter_subscriber = CounterSubscriber::new();
+        counter_subscriber.trigger_event();
+        event_manager.add_subscriber(counter_subscriber);
+    }
+
+    c.bench_function("process_basic", |b| {
+        b.iter(|| {
+            let ev_count = event_manager.run().unwrap();
+            assert_eq!(ev_count, 200)
+        })
+    });
+}
+
+pub fn run_arc_mutex_subscriber(c: &mut Criterion) {
+    let mut event_manager = EventManager::<Arc<Mutex<CounterSubscriber>>>::new().unwrap();
+    for _ in 0..200 {
+        let counter_subscriber = Arc::new(Mutex::new(CounterSubscriber::new()));
+        counter_subscriber.lock().unwrap().trigger_event();
+        event_manager.add_subscriber(counter_subscriber);
+    }
+
+    c.bench_function("process_with_arc_mutex", |b| {
+        b.iter(|| {
+            let ev_count = event_manager.run().unwrap();
+            assert_eq!(ev_count, 200)
+        })
+    });
+}
+
+pub fn run_subscriber_with_inner_mut(c: &mut Criterion) {
+    let mut event_manager = EventManager::<Arc<dyn EventSubscriber + Send + Sync>>::new().unwrap();
+    for _ in 0..200 {
+        let counter_subscriber = CounterInnerMutSubscriber::new();
+        counter_subscriber.trigger_event();
+        event_manager.add_subscriber(Arc::new(counter_subscriber));
+    }
+
+    c.bench_function("process_with_inner_mut", |b| {
+        b.iter(|| {
+            let ev_count = event_manager.run().unwrap();
+            assert_eq!(ev_count, 200)
+        })
+    });
+}
+
+pub fn run_multiple_subscriber_types(c: &mut Criterion) {
+    let mut event_manager = EventManager::<Arc<Mutex<dyn MutEventSubscriber>>>::new()
+        .expect("Cannot create event manager.");
+
+    for i in 0..100 {
+        // The `CounterSubscriberWithData` expects to receive a number as a parameter that represents
+        // the number it can use as its inner Events data.
+        let mut data_subscriber = CounterSubscriberWithData::new(i * 100);
+        data_subscriber.trigger_all_counters();
+        event_manager.add_subscriber(Arc::new(Mutex::new(data_subscriber)));
+
+        let mut counter_subscriber = CounterSubscriber::new();
+        counter_subscriber.trigger_event();
+        event_manager.add_subscriber(Arc::new(Mutex::new(counter_subscriber)));
+    }
+
+    c.bench_function("process_dynamic_dispatch", |b| {
+        b.iter(|| {
+            let _ = event_manager.run().unwrap();
+        })
+    });
+}
+
+pub fn run_subscriber_with_multiple_events(c: &mut Criterion) {
```

Comment:
Hmm, is this function testing something significantly different from the others? It looks like the difference is that `run` returns 256 events here vs 200 before.

Reply:
That's obviously not the only difference :rofl: 

It is testing the impact of using user specified data when registering events. It also checks the impact of having multiple ready events from the same subscriber.

I should add a comment to better explain how this test is different.

Reply:
I ended up deleting this test.